File: MtxVec.Delphi/Optim_TR.pas
unit Optim_TR;

interface

{$I BdsppDefs.inc}

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, Basic2, StdCtrls, ComCtrls, ExtCtrls, MtxVec, Types, MtxExpr, Optimization,
  Math387, TeEngine, Series, TeeProcs, Chart;

type
  TfrmTRDemo = class(TBasicForm2)
    RadioGroup1: TRadioGroup;
    Button1: TButton;
    Chart1: TChart;
    Series1: TBarSeries;
    Series2: TBarSeries;
    GroupBox1: TGroupBox;
    Label1: TLabel;
    Edit1: TEdit;
    Label2: TLabel;
    Edit2: TEdit;
    procedure FormCreate(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure RadioGroup1Click(Sender: TObject);                           
  private
    { Private declarations }
    IterArray: Array[0..6] of Integer;
    TimeArray: Array[0..6] of double;
    initpars: Array[0..1] of TSample;
    procedure TRTest;
    procedure SimplexTest;
    procedure MarquardtTest;
    procedure BFGSTest;
    procedure DFPTest;
    procedure ConjGradFletcher;
    procedure ConjGradPolack;
  public
    { Public declarations }
  end;

var
  frmTRDemo: TfrmTRDemo;

implementation

uses MtxIntDiff, AbstractMtxVec;

const NumRep = 200;

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

// initial estimates: (EPS,EPS)
procedure BananaVector(const x,f: TVec; const c: Array of double; const o: Array of TObject);
begin
  f[0] := 100*Sqr(x[1]-Sqr(x[0]));
  f[1] := Sqr(1-x[0]);
end;

// initial estimates: (EPS,EPS)
function BananaScalar(const x: TVec;Const c :TVec; Const o: Array of TObject): double;
begin
  Result := 100*Sqr(x.Values[1]-Sqr(x.Values[0]))+Sqr(1-x.Values[0]);
end;

procedure TfrmTRDemo.BFGSTest;
var
  i: Integer;
  x: Array[0..1] of double;
  stopReason: TOptStopReason;
  fMin: double;
  iHess: Matrix;
begin
  Math387.StartTimer;
  IterArray[3] := 0;
  for I := 1 to NumRep do
  begin
    x[0] := initpars[0];
    x[1] := initpars[1];
    IterArray[3] := IterArray[3] + BFGS(BananaScalar,NumericGradDifference,x,[],[],fMin,iHEss,stopREason,mvDouble,false,true,1000);
  end;
  TimeArray[3] := Math387.StopTimer*1000;
  IterArray[3] := Round(IterArray[3]/NumRep);
  if (stopReason=optResConverged) then
  begin
    Series1.Add(IterArray[3],'BFGS'+#13+'converged');
    Series2.Add(TimeArray[3],'BFGS'+#13+'converged');
  end
  else
  begin
    Series1.Add(IterArray[3],'BFGS'+#13+'NOT converged');
    Series2.Add(TimeArray[3],'BFGS'+#13+'NOT converged');
  end;
end;

procedure TfrmTRDemo.Button1Click(Sender: TObject);
begin
  Series1.Clear;
  Series2.Clear;
  initpars[0] := StrToSample(Edit1.Text);
  initpars[1] := StrToSample(Edit2.Text);
  Screen.Cursor := crHourGlass;
  try
    TRTest;
    SimplexTest;
    MarquardtTest;
    BFGSTest;
    DFPTest;
    ConjGradFletcher;
    ConjGradPolack;
  finally
    Screen.Cursor := crDefault;
  end;
  RadioGroup1Click(Self);
end;

procedure TfrmTRDemo.ConjGradFletcher;
var
  i: Integer;
  x: Array[0..1] of double;
  stopReason: TOptStopReason;
  fMin: double;
begin
  Math387.StartTimer;
  IterArray[5] := 0;
  for I := 1 to NumRep do
  begin
    x[0] := initpars[0];
    x[1] := initpars[1];
    IterArray[5] := IterArray[5] + ConjGrad(BananaScalar, NumericGradDifference,x,[],[],fMin,stopREason,mvDouble,true,true,1500);
  end;
  TimeArray[5] := StopTimer*1000;
  IterArray[5] := Round(IterArray[5]/NumRep);
  if (stopReason=optResConverged) then
  begin
    Series1.Add(IterArray[5],'Fletcher'+#13+'converged');
    Series2.Add(TimeArray[5],'Fletcher'+#13+'converged');
  end
  else
  begin
    Series1.Add(IterArray[5],'Fletcher'+#13+'NOT converged');
    Series2.Add(TimeArray[5],'Fletcher'+#13+'NOT converged');
  end;
end;

procedure TfrmTRDemo.ConjGradPolack;
var
  i: Integer;
  x: Array[0..1] of double;
  stopReason: TOptStopReason;
  fMin: double;
begin
  Math387.StartTimer;
  IterArray[6] := 0;
  for I := 1 to NumRep do
  begin
    x[0] := initpars[0];
    x[1] := initpars[1];
    IterArray[6] := IterArray[6] + ConjGrad(BananaScalar,NumericGradDifference,x,[],[],fMin,stopREason,mvDouble,false,true,1500);
  end;
  TimeArray[6] := StopTimer*1000;
  IterArray[6] := Round(IterArray[6]/NumRep);
  if (stopReason=optResConverged) then
  begin
    Series1.Add(IterArray[6],'Polack'+#13+'converged');
    Series2.Add(TimeArray[6],'Polack'+#13+'converged');
  end
  else
  begin
    Series1.Add(IterArray[6],'Polack'+#13+'NOT converged');
    Series2.Add(TimeArray[6],'Polack'+#13+'NOT converged');
  end;
end;

procedure TfrmTRDemo.DFPTest;
var
  i: Integer;
  x: Array[0..1] of double;
  stopReason: TOptStopReason;
  fMin: double;
  iHess: Matrix;
begin
  Math387.StartTimer;
  IterArray[4] := 0;
  for I := 1 to NumRep do
  begin
    x[0] := initpars[0];
    x[1] := initpars[1];
    IterArray[4] := IterArray[4] + BFGS(BananaScalar,NumericGradDifference,x,[],[],fMin,iHEss,stopREason,mvDouble,true);
  end;
  TimeArray[4] := StopTimer*1000;
  IterArray[4] := Round(IterArray[4]/NumRep);
  if (stopReason=optResConverged) then
  begin
    Series1.Add(IterArray[3],'DFP'+#13+'converged');
    Series2.Add(TimeArray[3],'DFP'+#13+'converged');
  end
  else
  begin
    Series1.Add(IterArray[3],'DFP'+#13+'NOT converged');
    Series2.Add(TimeArray[3],'DFP'+#13+'NOT converged');
  end;
end;

procedure TfrmTRDemo.FormCreate(Sender: TObject);
begin
  inherited;
  With RichEdit1.Lines do
  begin
    Clear;
    Add('New in MtxVec 3.0 : Trust Region optimization algorithm.');
    Add('Using TR algorithm it''s now possible to minimize vector function of '
      + ' several variables with or without bounds.');
    Add('');
    Add('This example minimizes the famous "Banana" function using several different '
      + ' algorithms and compares the results. As seen from results, TR algorithm is not '
      + 'the best choice if you''re minimizing small scale problems. TR algorithm is the '
      + 'preffered choice if you''re working on large scale bounded or unbouded problems.');
    Add('');
    Add('NOTE: Check http://plato.la.asu.edu/guide.html when in doubt which algorithm to use.');
  end;

  Edit1.Text := FormatSample(0.0);
  Edit2.Text := FormatSample(0.0);
end;

procedure TfrmTRDemo.MarquardtTest;
var x: Array[0..1] of double;
  StopReason: TOptStopReason;
  fmin: double;
  iHess: TMtx;
  i: Integer;
begin
  CreateIt(iHess);
  try
    Math387.StartTimer;
    IterArray[2] := 0;
    for I := 1 to NumRep do
    begin
      x[0] := initpars[0];
      x[1] := initpars[1];
      IterArray[2] := IterArray[2] + Marquardt(BananaScalar,NumericGradHess,x,[],[],fMin,iHess,StopReason,mvDouble,1000);
    end;
    TimeArray[2] := StopTimer*1000;
    IterArray[2] := Round(IterArray[2]/NumRep);
    if (stopReason=optResConverged) then
    begin
      Series1.Add(IterArray[2],'Marquardt'+#13+'converged');
      Series2.Add(TimeArray[2],'Marquardt'+#13+'converged');
    end
    else
    begin
      Series1.Add(IterArray[2],'Marquardt'+#13+'NOT converged');
      Series2.Add(TimeArray[2],'Marquardt'+#13+'NOT converged');
    end;
  finally
    FreeIt(iHess);
  end;
end;

procedure TfrmTRDemo.RadioGroup1Click(Sender: TObject);
begin
  Series1.Active := RadioGroup1.ItemIndex = 0;
  Series2.Active := RadioGroup1.ItemIndex = 1;
  Chart1.Title.Text.Clear;
  Case RadioGroup1.ItemIndex of
    0: Chart1.Title.Text.Add('Average number of iterations needed to converge.');
    1: Chart1.Title.Text.Add('Time needed to converge ('+ IntToStr(NumRep)+' iterations).');
  End;
end;

procedure TfrmTRDemo.SimplexTest;
var x: Array[0..1] of double;
  StopReason: TOptStopReason;
  fmin: double;
  i: Integer;
begin
  Math387.StartTimer;
  IterArray[1] := 0;
  for I := 1 to NumRep do
  begin
    x[0] := initpars[0];
    x[1] := initpars[1];
    IterArray[1] := IterArray[1] + Simplex(BananaScalar,x,[],[],fMin,StopReason,mvDouble,1000);
  end;
  TimeArray[1] := StopTimer*1000;
  IterArray[1] := Round(IterArray[1]/NumRep);
  if (stopReason=optResConverged) then
  begin
    Series1.Add(IterArray[1],'Simplex'+#13+'converged');
    Series2.Add(TimeArray[1],'Simplex'+#13+'converged');
  end
  else
  begin
    Series1.Add(IterArray[1],'Simplex'+#13+'NOT converged');
    Series2.Add(TimeArray[1],'Simplex'+#13+'NOT converged');
  end;
end;

procedure TfrmTRDemo.TRTest;
var x,y: Vector;
  EPSArray: TEPSArray;
  StopReason: TOptStopReason;
  i: Integer;
begin
  x.SetIt(false,[initpars[0],initpars[1]]);
  y.Size(2);

  {$IFDEF TTDOUBLE}
  EPSArray[0] := 1.0E-10;
  EPSArray[1] := 1.0E-8;
  EPSArray[2] := 1.0E-8;
  EPSArray[3] := 1.0E-8;
  EPSArray[4] := 1.0E-8;
  EPSArray[5] := 1.0E-10;
  {$ELSE}
  EPSArray[0] := 1.0E-5;
  EPSArray[1] := 1.0E-4;
  EPSArray[2] := 1.0E-4;
  EPSArray[3] := 1.0E-4;
  EPSArray[4] := 1.0E-4;
  EPSArray[5] := 1.0E-5;
  {$ENDIF}

  Math387.StartTimer;
  IterArray[0] := 0;
  for I := 1 to NumRep do
  begin
    x.SetIt(false,[initpars[0],initpars[1]]);
    IterArray[0] := IterArray[0] + TrustRegion(BananaVector,nil,x,y,[],[],1000,100,EPSArray,0.0,stopReason);
  end;
  TimeArray[0] := StopTimer*1000;
  IterArray[0] := Round(IterArray[0]/NumRep);
  if (stopReason=optResConverged) then
  begin
    Series1.Add(IterArray[0],'TR'+#13+'converged');
    Series2.Add(TimeArray[0],'TR'+#13+'converged');
  end
  else
  begin
    Series1.Add(IterArray[0],'TR'+#13+'NOT converged: ' + StopReasonToStr(stopReason));
    Series2.Add(TimeArray[0],'TR'+#13+'NOT converged: ' + StopReasonToStr(stopReason));
  end;
end;

initialization
  RegisterClass(TfrmTRDemo);
end.

File: MtxVec.CPP/Optim_TR.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "Optim_TR.h"
#include <MtxIntDiff.hpp>
#include <AbstractMtxVec.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic2"
#pragma resource "*.dfm"
TfrmTRDemo *frmTRDemo;
//---------------------------------------------------------------------------
static const int NumRep = 200;
//---------------------------------------------------------------------------
__fastcall TfrmTRDemo::TfrmTRDemo(TComponent* Owner)
  : TBasicForm2(Owner)
{
}
//---------------------------------------------------------------------------
static void __fastcall BananaVector(TVec * const x, TVec * const f,
                                     const double *c,
                                     #if (__BORLANDC__ >= 0x0770)
                                     const System::NativeInt c_High,
                                     #else
                                     const int c_Size,
                                     #endif
                                     System::TObject * const *o,
                                     #if (__BORLANDC__ >= 0x0770)
                                     const System::NativeInt o_High
                                     #else
                                     const int o_Size
                                     #endif
                                     )
{
  // Typecast TVec* -> TVector* (layout-identical subclass, MtxVec.h): its Values[]
  // property is inline raw pointer math (((double*)ValuesPointer)[i]) that shadows
  // the inherited DynamicArray Values FIELD -- fast, and correct on the SetSubRange
  // views the optimizer passes. Indexing the field directly (TVec::Values[i], or the
  // get_DefaultArray property) is wrong here: the field's C++ operator[] bounds-checks
  // against a counterfeit length and throws. See CPP-toolchain.md. Matches Delphi x[i]/f[i].
  Mtxvec::TVector *px = (Mtxvec::TVector*) x;
  Mtxvec::TVector *pf = (Mtxvec::TVector*) f;
  double x0 = px->Values[0];
  double x1 = px->Values[1];
  double t  = x1 - x0 * x0;
  double u  = 1 - x0;
  pf->Values[0] = 100 * t * t;
  pf->Values[1] = u * u;
}
//---------------------------------------------------------------------------
static double __fastcall BananaScalar(TVec * const x, TVec * const c,
                                       System::TObject * const *o,
                                       #if (__BORLANDC__ >= 0x0770)
                                       const System::NativeInt o_High
                                       #else
                                       const int o_Size
                                       #endif
                                       )
{
  // Typecast TVec* -> TVector* (MtxVec.h): Values[] = inline raw pointer math, correct
  // on the SetSubRange views Simplex passes. Indexing the inherited DynamicArray field
  // (TVec::Values[i]) would bounds-check against a counterfeit length and throw. See
  // CPP-toolchain.md "the Values[] field bounds-check trap". Matches Delphi.
  Mtxvec::TVector *px = (Mtxvec::TVector*) x;
  double x0 = px->Values[0];
  double x1 = px->Values[1];
  double t  = x1 - x0 * x0;
  double u  = 1 - x0;
  return 100 * t * t + u * u;
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::FormCreate(TObject *Sender)
{
  RichEdit1->Lines->Clear();
  RichEdit1->Lines->Add("New in MtxVec 3.0 : Trust Region optimization algorithm.");
  RichEdit1->Lines->Add(
    "Using TR algorithm it's now possible to minimize vector function of "
    " several variables with or without bounds.");
  RichEdit1->Lines->Add("");
  RichEdit1->Lines->Add(
    "This example minimizes the famous \"Banana\" function using several different "
    " algorithms and compares the results. As seen from results, TR algorithm is not "
    "the best choice if you're minimizing small scale problems. TR algorithm is the "
    "preffered choice if you're working on large scale bounded or unbouded problems.");
  RichEdit1->Lines->Add("");
  RichEdit1->Lines->Add("NOTE: Check http://plato.la.asu.edu/guide.html when in doubt which algorithm to use.");

  Edit1->Text = FormatSample(0.0);
  Edit2->Text = FormatSample(0.0);
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::Button1Click(TObject *Sender)
{
  Series1->Clear();
  Series2->Clear();
  initpars[0] = StrToSample(Edit1->Text);
  initpars[1] = StrToSample(Edit2->Text);
  Screen->Cursor = crHourGlass;
  try {
    TRTest();
    SimplexTest();
    MarquardtTest();
    BFGSTest();
    DFPTest();
    ConjGradFletcher();
    ConjGradPolack();
  } __finally {
    Screen->Cursor = crDefault;
  }
  RadioGroup1Click(this);
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::RadioGroup1Click(TObject *Sender)
{
  Series1->Active = RadioGroup1->ItemIndex == 0;
  Series2->Active = RadioGroup1->ItemIndex == 1;
  Chart1->Title->Text->Clear();
  switch (RadioGroup1->ItemIndex)
  {
    case 0: Chart1->Title->Text->Add("Average number of iterations needed to converge."); break;
    case 1: Chart1->Title->Text->Add("Time needed to converge (" + IntToStr(NumRep) + " iterations)."); break;
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::TRTest()
{
  sVector x, y;
  x.SetIt(false, OPENARRAY(double, (initpars[0], initpars[1])));
  y.Size(2);
  TEPSArray EPSArray;
  EPSArray[0] = 1.0E-10;
  EPSArray[1] = 1.0E-8;
  EPSArray[2] = 1.0E-8;
  EPSArray[3] = 1.0E-8;
  EPSArray[4] = 1.0E-8;
  EPSArray[5] = 1.0E-10;

  StartTimer();
  IterArray[0] = 0;
  TOptStopReason stopReason;
  for (int i = 1; i <= NumRep; i++)
  {
    x.SetIt(false, OPENARRAY(double, (initpars[0], initpars[1])));
    IterArray[0] += TrustRegion(BananaVector, NULL, x, y, NULL, -1, NULL, -1, 1000, 100, EPSArray, 0.0, stopReason);
  }
  TimeArray[0] = StopTimer() * 1000;
  IterArray[0] = Math387::RoundToInt((double)IterArray[0] / NumRep);
  if (stopReason == OptResConverged)
  {
    Series1->Add(IterArray[0], "TR\rconverged");
    Series2->Add(TimeArray[0], "TR\rconverged");
  }
  else
  {
    Series1->Add(IterArray[0], "TR\rNOT converged: " + StopReasonToStr(stopReason));
    Series2->Add(TimeArray[0], "TR\rNOT converged: " + StopReasonToStr(stopReason));
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::SimplexTest()
{
  double x[2];
  StartTimer();
  IterArray[1] = 0;
  TOptStopReason stopReason;
  double fMin;
  for (int i = 1; i <= NumRep; i++)
  {
    x[0] = initpars[0];
    x[1] = initpars[1];
    IterArray[1] += Simplex(BananaScalar, &x[0], 1, NULL, -1, NULL, -1, fMin, stopReason, mvDouble, 1000);
  }
  TimeArray[1] = StopTimer() * 1000;
  IterArray[1] = Math387::RoundToInt((double)IterArray[1] / NumRep);
  AnsiString lbl = (stopReason == OptResConverged) ? "Simplex\rconverged" : "Simplex\rNOT converged";
  Series1->Add(IterArray[1], lbl);
  Series2->Add(TimeArray[1], lbl);
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::MarquardtTest()
{
  double x[2];
  TMtx *iHess;
  CreateIt(iHess);
  try {
    StartTimer();
    IterArray[2] = 0;
    TOptStopReason stopReason;
    double fMin;
    for (int i = 1; i <= NumRep; i++)
    {
      x[0] = initpars[0];
      x[1] = initpars[1];
      IterArray[2] += Marquardt(BananaScalar, NumericGradHess, &x[0], 1, NULL, -1, NULL, -1,
                                fMin, iHess, stopReason, mvDouble, 1000);
    }
    TimeArray[2] = StopTimer() * 1000;
    IterArray[2] = Math387::RoundToInt((double)IterArray[2] / NumRep);
    AnsiString lbl = (stopReason == OptResConverged) ? "Marquardt\rconverged" : "Marquardt\rNOT converged";
    Series1->Add(IterArray[2], lbl);
    Series2->Add(TimeArray[2], lbl);
  } __finally {
    FreeIt(iHess);
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::BFGSTest()
{
  double x[2];
  sMatrix iHess;
  StartTimer();
  IterArray[3] = 0;
  TOptStopReason stopReason;
  double fMin;
  for (int i = 1; i <= NumRep; i++)
  {
    x[0] = initpars[0];
    x[1] = initpars[1];
    IterArray[3] += BFGS(BananaScalar, NumericGradDifference, &x[0], 1, NULL, -1, NULL, -1,
                         fMin, iHess, stopReason, mvDouble, false, true, 1000);
  }
  TimeArray[3] = StopTimer() * 1000;
  IterArray[3] = Math387::RoundToInt((double)IterArray[3] / NumRep);
  AnsiString lbl = (stopReason == OptResConverged) ? "BFGS\rconverged" : "BFGS\rNOT converged";
  Series1->Add(IterArray[3], lbl);
  Series2->Add(TimeArray[3], lbl);
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::DFPTest()
{
  double x[2];
  sMatrix iHess;
  StartTimer();
  IterArray[4] = 0;
  TOptStopReason stopReason;
  double fMin;
  for (int i = 1; i <= NumRep; i++)
  {
    x[0] = initpars[0];
    x[1] = initpars[1];
    IterArray[4] += BFGS(BananaScalar, NumericGradDifference, &x[0], 1, NULL, -1, NULL, -1,
                         fMin, mvDouble, true);
  }
  TimeArray[4] = StopTimer() * 1000;
  IterArray[4] = Math387::RoundToInt((double)IterArray[4] / NumRep);
  AnsiString lbl = (stopReason == OptResConverged) ? "DFP\rconverged" : "DFP\rNOT converged";
  Series1->Add(IterArray[3], lbl);
  Series2->Add(TimeArray[3], lbl);
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::ConjGradFletcher()
{
  double x[2];
  StartTimer();
  IterArray[5] = 0;
  TOptStopReason stopReason;
  double fMin;
  for (int i = 1; i <= NumRep; i++)
  {
    x[0] = initpars[0];
    x[1] = initpars[1];
    IterArray[5] += ConjGrad(BananaScalar, NumericGradDifference, &x[0], 1, NULL, -1, NULL, -1,
                              fMin, stopReason, mvDouble, true, true, 1500);
  }
  TimeArray[5] = StopTimer() * 1000;
  IterArray[5] = Math387::RoundToInt((double)IterArray[5] / NumRep);
  AnsiString lbl = (stopReason == OptResConverged) ? "Fletcher\rconverged" : "Fletcher\rNOT converged";
  Series1->Add(IterArray[5], lbl);
  Series2->Add(TimeArray[5], lbl);
}
//---------------------------------------------------------------------------
void __fastcall TfrmTRDemo::ConjGradPolack()
{
  double x[2];
  StartTimer();
  IterArray[6] = 0;
  TOptStopReason stopReason;
  double fMin;
  for (int i = 1; i <= NumRep; i++)
  {
    x[0] = initpars[0];
    x[1] = initpars[1];
    IterArray[6] += ConjGrad(BananaScalar, NumericGradDifference, &x[0], 1, NULL, -1, NULL, -1,
                              fMin, stopReason, mvDouble, false, true, 1500);
  }
  TimeArray[6] = StopTimer() * 1000;
  IterArray[6] = Math387::RoundToInt((double)IterArray[6] / NumRep);
  AnsiString lbl = (stopReason == OptResConverged) ? "Polack\rconverged" : "Polack\rNOT converged";
  Series1->Add(IterArray[6], lbl);
  Series2->Add(TimeArray[6], lbl);
}
//---------------------------------------------------------------------------
